src/main/java/org/folio/rest/impl/InvoiceLineHelper.java

```diff
@@ -276,19 +308,244 @@ private String getInvoiceLineNumberEndpoint(String id) {
+
+      // Re-apply prorated adjustments
+      applyProratedAdjustments(proratedAdjustments, allLines, invoice);
+
+      // Return only those related lines which were updated after re-applying prorated adjustment(s)
+      return lines.stream()
+        .filter(line -> !originalLineHashCodes.contains(line.hashCode()))
+        .collect(toList());
+    });
   }
 
-  private void updateInvoice(String invoiceId) {
+  /**
+   * Gets all other invoice lines associated with the same invoice. Passed {@code invoiceLine} is not added to the result.
+   *
+   * @param invoiceLine {@link InvoiceLine} record
+   * @return list of all other invoice lines associated with the same invoice
+   */
+  private CompletableFuture<List<InvoiceLine>> getRelatedLines(InvoiceLine invoiceLine) {
+    String cql = String.format(QUERY_BY_INVOICE_ID, invoiceLine.getInvoiceId());
+    if (invoiceLine.getId() != null) {
+      cql = combineCqlExpressions("and", cql, "id<>" + invoiceLine.getId());
+    }
+    String endpoint = String.format(GET_INVOICE_LINES_BY_QUERY, Integer.MAX_VALUE, 0, getEndpointWithQuery(cql, logger), lang);
+
+    return getInvoiceLineCollection(endpoint).thenApply(InvoiceLineCollection::getInvoiceLines);
+  }
+
+  private void updateInvoiceAndAffectedLinesAsync(Invoice invoice, List<InvoiceLine> lines) {
+    VertxCompletableFuture.runAsync(ctx, () -> {
+      InvoiceLineHelper helper = new InvoiceLineHelper(okapiHeaders, ctx, lang);
+      helper.persistInvoiceLines(invoice, lines).handle((ok, fail) -> {
+        if (fail == null) {
+          updateInvoiceAsync(invoice);
+        }
+        helper.closeHttpClient();
+        return null;
+      });
+    });
+  }
+
+  private CompletableFuture<Void> persistInvoiceLines(Invoice invoice, List<InvoiceLine> lines) {
+    return VertxCompletableFuture.allOf(ctx, lines.stream()
+      .map(invoiceLine -> {
+        calculateInvoiceLineTotals(invoiceLine, invoice);
+        return this.updateInvoiceLineToStorage(invoiceLine);
+      })
+      .toArray(CompletableFuture[]::new));
+  }
+
+  private void updateInvoiceAsync(Invoice invoice) {
     VertxCompletableFuture.runAsync(ctx,
-        () -> sendEvent(MessageAddress.INVOICE_TOTALS, new JsonObject().put(INVOICE_ID, invoiceId)));
+      () -> sendEvent(MessageAddress.INVOICE_TOTALS, new JsonObject().put(INVOICE, JsonObject.mapFrom(invoice))));
+  }
+
+  private void updateInvoiceAndLinesAsync(String invoiceId) {
+    VertxCompletableFuture.runAsync(ctx, () -> {
+      InvoiceLineHelper helper = new InvoiceLineHelper(okapiHeaders, ctx, lang);
+      helper.updateInvoiceAndLines(invoiceId).handle((ok, fail) -> {
+        helper.closeHttpClient();
+        return null;
+      });
+    });
+  }
+
+  private CompletableFuture<Void> updateInvoiceAndLines(String invoiceId) {
+    return getInvoiceById(invoiceId, lang, httpClient, ctx, okapiHeaders, logger)
+      .thenCompose(invoice -> getInvoiceLinesByInvoiceId(invoiceId).thenApply(lines -> {
```

Comment:
As with the `applyProratedAdjustments(InvoiceLine invoiceLine, Invoice invoice)` method, there is no need to retrieve and recalculate the lines if there are no prorated adjustments in the invoice.

Reply:
Added extra condition